### DISCUSSION TITLE
Middleware Question

### DISCUSSION THREAD

**Discussion by Joseppi83** [vercel/next.js#48648] [Help]
### Summary

I just got back into Next.js, after spending about 3 years with C/C++/Java.  And a lot has changed, which is exciting, because it's got my mind on overdrive right now.

In any case, in response to the Middleware that was added, in the last hobby project I produced with Next, I merely added a middleware directory inside the api directory, then imported the middleware function inside the API route where it occured.  Or, in the case with checking for authentication and authorization for api routes, I used an asynchronous wrapper function for that route, so it was passed into an isUserAuthenticated or isUserAuthorized function first.

Is that also a valid option to run middleware?

**Comment by Joseppi83** [comment]
Wow.  Disregard.  There are so many extensions and libraries for Next now, as opposed to a few years ago.  Just became aware of next-auth.

**Comment by Hamzi-SE** [comment]
You can use [clerk](https://clerk.com/docs/nextjs/overview) for authentication/authorization also. It's pretty simple and straightforward.

**Comment by Joseppi83** [comment]
Wow.  Thanks, man.  I was completely unaware of that one also.  I already setup authentication/authorization manually a few years ago, but didn't completely add CSRF to some routes.

That aside, it seems like next-auth is probably the best route.  When I see people from Vercel mentioning it in YouTube videos, that likely means they might eventually just integrate it into the Next framework.  Plus, next-auth seems to have much more support, so I'll likely trash my manual auth code, refactor, and integrate the strongest backed auth compliant with next.

**Reply by Hamzi-SE** [reply]
Yeah I also like [next-auth](https://next-auth.js.org/) the most :)
